File: leetcode/2368-reachable-nodes-with-restrictions/solution.ts
function reachableNodes(
  n: number,
  edges: number[][],
  restricted: number[]
): number {
  const graph = Array.from({ length: n }, () => []);
  const restrictedSet = new Set(restricted);

  for (let i = 0; i < edges.length; i++) {
    if (restrictedSet.has(edges[i][0]) || restrictedSet.has(edges[i][1])) {
      continue;
    }

    graph[edges[i][0]].push(edges[i][1]);
    graph[edges[i][1]].push(edges[i][0]);
  }

  const seen = Array.from({ length: n }, () => false);

  function dfs(root: number) {
    seen[root] = true;
    let visited = 0;

    for (let i = 0; i < graph[root].length; i++) {
      const neighbour = graph[root][i];

      if (!seen[neighbour]) {
        visited += dfs(neighbour);
      }
    }

    return visited + 1;
  }

  return dfs(0);
}

File: leetcode/2368-reachable-nodes-with-restrictions/solution.cpp
int dfs(int root, vector<vector<int>>& graph, vector<bool>& seen) {
    seen[root] = true;
    int visited = 0;

    for (int i = 0; i < graph[root].size(); i++) {
        if (!seen[graph[root][i]]) {
            visited += dfs(graph[root][i], graph, seen);
        }
    }

    return visited + 1;
}

int reachableNodes(int n, vector<vector<int>>& edges, vector<int>& restricted) {
    vector<vector<int>> graph(n + 1, vector<int>());
    unordered_set<int> restricted_set(begin(restricted), end(restricted));

    for (int i = 0; i < edges.size(); i++) {
        if (restricted_set.find(edges[i][0]) != restricted_set.end() || restricted_set.find(edges[i][1]) != restricted_set.end()) {
            continue;
        }

        graph[edges[i][0]].push_back(edges[i][1]);
        graph[edges[i][1]].push_back(edges[i][0]);
    }

    vector<bool> seen(n, false);

    return dfs(0, graph, seen);
}